content/blog/r-frictionless-validation.md

```diff
@@ -0,0 +1,85 @@
+ 
+In this post we will look at two R packages for working with Frictionless Data Packages: `frictionless` and `datapackage.r`. Both are developed by Frictionless Data organisation. 
+
+## Frictionless
+
+`frictionless` is relatively new, available on CRAN, peer-reviewed by ROpenSci and under active development. However, note that the [package is not designed to offer "full validation"](https://github.com/frictionlessdata/frictionless-r/issues/125#issuecomment-1477422737) which will be somewhat limiting. It does have some checking against the table schema, so let's explore what is possible. Firstly the `frictionless` R package will only work with data stored on disk like
+
+```
+package_name
+├── datapackage.json
+├── file1.csv
+└── file2.csv
+```
+
+there is no option to supply a `datapackage.json` and load data from elsewhere.
+
+The package json contains a [table schema](https://specs.frictionlessdata.io/table-schema/) for each resource which allows specifying the columns, their `type` and any `constraints` on the column. One of the validation checks we want is to ensure that the types in a column are correct. If we try and read a resource using `frictionless` which has an invalid type we will get a warning like
+
+```R
+Warning message:                                                                                                                                                                      
+One or more parsing issues, call `problems()` on your data frame for details, e.g.:
+  dat <- vroom(...)
+  problems(dat) 
+```
+
+And running `vroom::problems()` we see
+
+```R
+> vroom::problems()
+# A tibble: 1 × 5
+     row   col expected actual file               
+   <int> <int> <chr>    <chr>  <chr>
+ 1     1    3 a double hello  /path/to/file.csv
+```
+
+We can see the row and column with the invalid type, we can also see the actual value and the expected type. This is useful information but it isn't a straightforward message we can surface to the end user. It will still require some work to build an error string from the returned information.
+
+Let's try adding a `constraint` to our table schema to ensure that a column's values match a set of possible values. If we load some invalid data using this schema, again we get a warning which if when we run `vroom::problems()` we can see the details of
+
+```R
+# A tibble: 1 × 5
+     row   col expected           actual   file 
+   <int> <int> <chr>              <chr>    <chr>
+ 1     1     1 value in level set other    ""  
+```
+
+Again `frictionless` tell us the offending row and column and the value found in that column, but it doesn't tell us what possible values it could have taken, making it less actionable. The end user would have to have knowledge of the schema to be able to fix this.
```

Comment:
```suggestion
Again `frictionless` tells us the offending row and column and the value found in that column, but it doesn't tell us what possible values it could have taken, making it less actionable. The end user would have to have knowledge of the schema to be able to fix this.
```